systemd/cloud-config.service.tmpl

```diff
@@ -10,7 +10,7 @@ ConditionEnvironment=!KERNEL_CMDLINE=cloud-init=disabled
 
 [Service]
 Type=oneshot
-ExecStart=/usr/bin/cloud-init modules --mode=config
+ExecStart=sh -c 'echo "start" | nc.openbsd -Uu -W1 /run/cloud-init/share/config.sock -s /run/cloud-init/share/config-return.sock | sh'
```

Comment:
> This is a fairly complex invocation for an ExecStart with no surrounding commentary or documentation to aid in explaining the intent of the socket setup and pipe chain.

Agreed, inline-documentation would help a lot. I'm happy to add that.

> It also makes typical triage entrypoints harder to read when folks may jump into systemctl status cloud-*.service as it pushes visibility of exit codes off screen because our ExecStart is so long.

You're referring to systemd's pager truncating the message on the `Process:` line, right?
```bash
$ systemctl status cloud-init-local.service
● cloud-init-local.service - Cloud-init: Local Stage (pre-network)
     Loaded: loaded (/usr/lib/systemd/system/cloud-init-local.service; enabled; preset: enabled)
     Active: active (exited) since Wed 2024-07-24 19:53:46 UTC; 2min 25s ago
    Process: 254 ExecStart=sh -c echo "start" | nc.openbsd -Uu -W1 /run/cloud-init/share/local.sock -s /run/cloud-init/share/local-return.sock | sh (code=exited, status=0/S>
   Main PID: 254 (code=exited, status=0/SUCCESS)
        CPU: 5ms
```

I don't think that this is really a concern. This off-screen information is duplicated on the following line, and even if it wasn't, one can always use `--no-pager` or pipe to `cat` / `less` / etc.

> It may aid in readability/maintainability if cloud-init delivered a simple wrapper script in /usr/lib/cloudinit/that would take a single command line param for the stage name (local|network|config|final) and create these seperate named sockets. We can then document in that script the intent of opening the socket and the act of passing through socket payloads to sh.

I'd really rather not hide the implementation in a script. The original implementation was dead simple - just a single `nc` invocation but at the request of getting error codes into the shim services I made this compromise to the previous simplicity. I'm very averse to making this any more complex than it already is. I don't want another `ds-identify`. Frankly I'd sooner drop the error codes in the shim services or re-write this one-liner in C/Rust/Go (but in preference of shipping code that we already have packaging for I've proposed it like this for now with the intent of potentially rewriting in a different language eventually).  

Also, putting this in a file would add a file load and read to the critical chain per-stage.

Reply:
> Agreed, inline-documentation would help a lot. I'm happy to add that.
This meets the majority of my concern, I'm good with your concerns about adding a wrapper layer adding complexity or obscurity where it may not be needed.

If we have comments within this unit that describe the intent of this interaction, then I can back down on the "please make systemctl status cloud-init.service more readable/simple"  point.

Reply:
See the latest commit, which I think addresses the remaining concerns in this PR.

Reply:
Yes looking great.